[Feature] Support more specific error message

stonechen111 opened the pull request:
1. Add a parameter to Starrocks::Status::to_string to control whether the status information it outputs has context.
2. When tablet creation fails, return the error message from be to the user client through Starrocks::Status::to_string.


## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

imay commented on be/src/common/status.h:
```diff
@@ -329,6 +332,10 @@ class STATUS_ATTRIBUTE Status {
 
+    /// @return A string similar to to_string(), but it is for the user client
+    /// and does not contain sensitive information.
+    std::string to_string_for_client() const;
```
please add unit test to cover this new function.

imay commented on be/src/agent/agent_task.cpp:
```diff
@@ -223,7 +223,7 @@ void run_create_tablet_task(const std::shared_ptr<CreateTabletAgentTaskRequest>&
         status_code = TStatusCode::RUNTIME_ERROR;
         if (tablet_type == TTabletType::TABLET_TYPE_LAKE) {
-            error_msgs.emplace_back("create tablet failed");
+            error_msgs.emplace_back(create_status.to_string_for_client());
```
Why not keep the "Fail to create tablet: " to let users know how this error message is generated?

stonechen111 replied:
The old error message contained "failed to create tablet" twice.
![image](https://github.com/StarRocks/starrocks/assets/145319350/18d44874-5340-4e16-afc5-f43f49cce164)

This "Fail to create tablet" is not necessary. It is replaced by a more detailed error message. The other "fail to create tablet:" is added in the following fucntion:
![image](https://github.com/StarRocks/starrocks/assets/145319350/1fd92346-0cf3-4af0-9f17-9798ed9cfedf)

With this commit, the error message to client now is like:
![image](https://github.com/StarRocks/starrocks/assets/145319350/9ab22205-4617-47d3-91c5-db758eaa3fa8)

imay commented on be/src/common/status.h:
```diff
@@ -399,6 +409,9 @@ class STATUS_ATTRIBUTE Status {
     const char* _state = nullptr;
+
+    // message to user clients that does not contain sensitive content
+    const char* _state_for_client = nullptr;
```
If you add a new member here, this Status structure will be doubled.
This will influence almost all the functions in StarRocks.
Could you please why this member was introduced and is there any other method to achieve it?

sduzh replied:
I'd prefer add a new function to remove context info from the Status, or create a new Status from another withou context info.

stonechen111 committed:
[Feature] Support more specific error message

Added Status::to_string_for_client method to report errors to users.
Errors reported by this method do not include file names and line numbers.

Signed-off-by: stonechen111 <chenshi@starrocks.com>

sduzh commented on be/src/agent/agent_task.cpp:
```diff
@@ -223,7 +223,8 @@ void run_create_tablet_task(const std::shared_ptr<CreateTabletAgentTaskRequest>&
         if (tablet_type == TTabletType::TABLET_TYPE_LAKE) {
-            error_msgs.emplace_back("create tablet failed");
+            bool with_context_info = false;
+            error_msgs.emplace_back(create_status.to_string(with_context_info));
```
```suggestion
            error_msgs.emplace_back(fmt::format("Fail to create tablet: {}", create_status.to_string(false)));
```

stonechen111 replied:
![image](https://github.com/StarRocks/starrocks/assets/145319350/b25dd660-2455-49fc-8bb0-9aed14d0a0b9)
这里会有"failed to create tablet:”的信息。当前的显示结果如下：
![image](https://github.com/StarRocks/starrocks/assets/145319350/1d772d22-d134-4dd6-82ab-2288865b581b)

应该不再需要加Fail to create tablet: {}了

imay commented on be/src/common/status.h:
```diff
@@ -327,7 +327,7 @@ class STATUS_ATTRIBUTE Status {
     /// @return A string representation of this status suitable for printing.
     ///   Returns the string "OK" for success.
-    std::string to_string() const;
+    std::string to_string(bool with_context_info = true) const;
```
Please add UT with with_context_info enabled and disabled.

stonechen111 replied:
OK

imay commented on be/src/agent/agent_task.cpp:
```diff
@@ -223,7 +223,8 @@ void run_create_tablet_task(const std::shared_ptr<CreateTabletAgentTaskRequest>&
         status_code = TStatusCode::RUNTIME_ERROR;
         if (tablet_type == TTabletType::TABLET_TYPE_LAKE) {
-            error_msgs.emplace_back("create tablet failed");
+            bool with_context_info = false;
+            error_msgs.emplace_back(create_status.to_string(with_context_info));
```
```suggestion
            error_msgs.emplace_back(create_status.to_string(fasle));
```

stonechen111 replied:
OK

kevincai reviewed (commented):
fix the pull request description as well, it is no longer valid for the actual code change.

stonechen111 committed:
add unit test

Signed-off-by: stonechen111 <chenshi@starrocks.com>

sduzh commented:
@Mergifyio backport branch-3.2

sduzh commented:
@Mergifyio backport branch-3.1
